### PR TITLE
[WIP] Fix build of Numpy with Intel

### LINKED ISSUE #3204: Unable to build Numpy with Intel 17
@davydden @glennpj @alalazo 

I'm trying to build `py-numpy` with Intel MKL, but although the build succeeds, numpy cannot be imported:
```
$ python -c 'import numpy'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.1/py-numpy-1.12.0-r7qapsmckwr6jjvb6llkoufk36g5t36h/lib/python2.7/site-packages/numpy-1.12.0-py2.7-linux-x86_64.egg/numpy/__init__.py", line 142, in <module>
    from . import add_newdocs
  File "/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.1/py-numpy-1.12.0-r7qapsmckwr6jjvb6llkoufk36g5t36h/lib/python2.7/site-packages/numpy-1.12.0-py2.7-linux-x86_64.egg/numpy/add_newdocs.py", line 13, in <module>
    from numpy.lib import add_newdoc
  File "/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.1/py-numpy-1.12.0-r7qapsmckwr6jjvb6llkoufk36g5t36h/lib/python2.7/site-packages/numpy-1.12.0-py2.7-linux-x86_64.egg/numpy/lib/__init__.py", line 8, in <module>
    from .type_check import *
  File "/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.1/py-numpy-1.12.0-r7qapsmckwr6jjvb6llkoufk36g5t36h/lib/python2.7/site-packages/numpy-1.12.0-py2.7-linux-x86_64.egg/numpy/lib/type_check.py", line 11, in <module>
    import numpy.core.numeric as _nx
  File "/blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/intel-17.0.1/py-numpy-1.12.0-r7qapsmckwr6jjvb6llkoufk36g5t36h/lib/python2.7/site-packages/numpy-1.12.0-py2.7-linux-x86_64.egg/numpy/core/__init__.py", line 24, in <module>
    raise ImportError(msg)
ImportError: 
Importing the multiarray numpy extension module failed.  Most
likely you are trying to import a failed build of numpy.
If you're working with a numpy git repo, try `git clean -xdf` (removes all
files not under version control).  Otherwise reinstall numpy.
```
Does anyone know how to get numpy + MKL working?

P.S. My import tests from #2869 will detect this type of failed build before the installation is complete. On the current develop, I didn't notice there was a problem with numpy until I tried building scipy.

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #3204.

Or at least, that's what I would like to say. I tried pretty hard to follow https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl but I'm still seeing the same error message. I opened an issue with the numpy developers (https://github.com/numpy/numpy/issues/8653). Hopefully they can get to the bottom of this.

I figured I would open up this WIP PR in the meantime to show what I'm trying to do. It offers several advantages over the way we were previously doing things.

If I'm going to rework this stuff, I want to do it right this time. Before this PR is merged, I plan on running the full numpy and scipy test suite for the following combinations:

- [ ] Intel 17 - Intel MKL - Python 2 - CentOS 6
- [ ] Intel 17 - Intel MKL - Python 3 - CentOS 6
- [ ] GCC 6 - OpenBLAS - CentOS 6
- [ ] GCC 6 - ATLAS - CentOS 6
- [ ] Clang 8 - OpenBLAS - macOS 10.12
- [ ] Clang 8 - Netlib BLAS/LAPACK - macOS 10.12

Any other combinations anyone would like to see?

Intel also provides a numpy benchmark. I may try that out with a few combinations and report back the performance. I'm curious as to how much of an improvement I get with Intel MKL.

### CODE REVIEW

**Comment by davydden** [comment]
> Any other combinations anyone would like to see?

Clang 8 - Openblas - macOS 10.12 (openblas is the default one, so it's good to test it).

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 1]
```diff
@@ -67,68 +70,93 @@ def setup_dependent_package(self, module, dep_spec):
...
-
-                f.write('library_dirs=%s\n' % dirs)
-                if not ((platform.system() == "Darwin") and
-                        (platform.mac_ver()[0] == '10.12')):
-                    f.write('rpath=%s\n' % dirs)
+                    f.write('[DEFAULT]\n')
+                f.write('libraries = {0}\n'.format(libraries))
+
+            include_dirs = os.pathsep.join(lapackblas_includes)
+            library_dirs = os.pathsep.join(lapackblas_libraries.directories)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: are you sure it's this instead of `':'.join(lapackblas.directories)`? That's what I saw being used in examples.

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 2]
```diff
@@ -67,68 +70,93 @@ def setup_dependent_package(self, module, dep_spec):
...
+                # library_dirs are in compilers_and_libraries_2017/linux/mkl/lib/intel64  # noqa
+                # include_dirs are in compilers_and_libraries_2017/linux/mkl/include      # noqa
+                lapackblas_includes = set()
+                for libdir in lapackblas_libraries.directories:
+                    incdir = os.path.dirname(os.path.dirname(libdir))
+                    incdir = join_path(incdir, 'include')
+                    lapackblas_includes.add(incdir)
+
+                f.write('[mkl]\n')
+                f.write('mkl_libs = {0}\n'.format('mkl_rt'))
```
Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: i would add some `FIXME` here as `mkl_rt` is really a hack because `numpy` does not configure with separately specified  threading and interface layers. Not really a Spack's fault.

File: var/spack/repos/builtin/packages/py-numpy/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: `os.pathsep` _is_ `:`, on most Unix platforms:
https://docs.python.org/2/library/os.html#os.pathsep

Apparently some use `,` though?? I decided to make this change because that's what numpy specifically recommended:
https://github.com/numpy/numpy/blob/master/site.cfg.example#L27

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: oh, i see.

File: var/spack/repos/builtin/packages/py-numpy/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: I'll have to scroll back through previous debates, but is it really that hacky? It's what Intel explicitly recommends:
https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: the point is that `mkl_rt` completely ingores any variants we use in `mkl` such as 32bit or 64bit, single threaded or multithreaded. So you can think of `mkl_rt` as an easy-to-make-it-work approach, but in Spack it's something we decided not to use to be rigorous about which version of Blas/Lapack is used within the DAG. So yes, it's hacky from Spack's perspective. Most likely the reasons why Intel recommends it is because `numpy` does NOT work otherwise, although i should!

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: here's an example of limitations in `numpy`'s config system https://github.com/numpy/numpy/issues/8294 -- you set `site.cfg` to use single-threaded Blas `satlas`, but it picks up multithreaded one.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Ok, I can add a comment.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Done.

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 3]
```diff
@@ -67,68 +70,105 @@ def setup_dependent_package(self, module, dep_spec):
...
+            include_dirs = os.pathsep.join(lapackblas_includes)
+            library_dirs = os.pathsep.join(lapackblas_libraries.directories)
+
+            f.write('include_dirs = {0}\n'.format(include_dirs))
+            f.write('library_dirs = {0}\n'.format(library_dirs))
+
+            # macOS 10.12 does not support RPATHs
+            if not ((platform.system() == "Darwin") and
+                    (platform.mac_ver()[0] == '10.12')):
+                f.write('rpath = {0}\n'.format(library_dirs))
```
Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: so you actually moved `rpath` here from `[default]` to each blas/lapack. I think their build config does not like it.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Actually, it was never in `[ALL]`/`[DEFAULT]` to begin with. That section wasn't (and still isn't) printed unless we are using some BLAS/LAPACK aside from openblas, mkl, and atlas.

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: yes, that's correct. I was mistaken.

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: @adamjstewart nope, i am not! here you go https://github.com/LLNL/spack/commit/f71df7bda3c4364ba6acd20411e8acac1c8f754c it was moved. I am not crazy after all.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Only slightly crazy. That was in #3028, not this PR.

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: The point is that

> Actually, it was never in [ALL]/[DEFAULT] to begin with. 

is not correct. It **was** there when I tried building `numpy` in https://github.com/LLNL/spack/pull/2365 . I just forgot that it was moved in https://github.com/LLNL/spack/pull/3028. But apparently the whole thing is still broken, so it does not matter at all.

**Comment by adamjstewart** [comment]
Since setting `rpath` in `site.cfg` doesn't seem to work, I wonder if we can manually append the correct directories to Spack's compiler wrapers. @tgamblin?

**Comment by davydden** [comment]
>  I wonder if we can manually append the correct directories to Spack's compiler wrapers

i don't think it's possible currently.`SPACK_RPATH_DEPS` is set in https://github.com/LLNL/spack/blob/develop/lib/spack/spack/build_environment.py#L271-L275 and then compiler wrappers rework them in https://github.com/LLNL/spack/blob/develop/lib/spack/env/cc#L278-L314

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 4]
```diff
@@ -67,68 +72,105 @@ def setup_dependent_package(self, module, dep_spec):
...
         args = []
 
         # From NumPy 1.10.0 on it's possible to do a parallel build
         if self.version >= Version('1.10.0'):
-            args = ['-j', str(make_jobs)]
+            args.extend(['-j', str(make_jobs)])
+
+        if self.compiler.name == 'intel':
+            if spec.satisfies('target=x86_64'):
+                args.extend(['--compiler=intelem', '--fcompiler=intelem'])
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Whether or not I add these flags seems to make a big difference. If I don't supply them, it still uses `icc`, but I think it is passing GCC flags. If I use these flags, the build crashes when it can't find `xiar` because it isn't in my PATH.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Also, when I don't use these flags and run `spack install py-numpy +blas ^openblas`, I get a broken build that correctly links to openblas. But when I add the flags and load the Intel parallel studio module, I get a working build that links to Intel MKL. Sigh...

Comment (davydden) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: maybe add a brief incode documentation of this fact?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: I'll see if I can figure out what's going on and potentially fix it first. Really wish some numpy devs would get back to me...

**Comment by adamjstewart** [comment]
Closing as I never got this working.
